api/src/main/java/io/minio/credentials/IamAwsProvider.java

```diff
@@ -0,0 +1,218 @@
+ * for Amazon EC2</a>.
+ */
+public class IamAwsProvider extends EnvironmentProvider {
+  // Custom endpoint to fetch IAM role credentials.
+  private final HttpUrl customEndpoint;
+  private final OkHttpClient httpClient;
+  private final ObjectMapper mapper;
+  private Credentials credentials;
+
+  public IamAwsProvider(@Nullable String customEndpoint, @Nullable OkHttpClient customHttpClient) {
+    this.customEndpoint =
+        (customEndpoint != null)
+            ? Objects.requireNonNull(HttpUrl.parse(customEndpoint), "Invalid custom endpoint")
+            : null;
+    // HTTP/1.1 is only supported in default client because of HTTP/2 in OkHttpClient cause 5
+    // minutes timeout on program exit.
+    this.httpClient =
+        (customHttpClient != null)
+            ? customHttpClient
+            : new OkHttpClient().newBuilder().protocols(Arrays.asList(Protocol.HTTP_1_1)).build();
+    this.mapper = new ObjectMapper();
+    this.mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    this.mapper.configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true);
+  }
+
+  private void checkLoopbackHost(HttpUrl url) {
+    try {
+      for (InetAddress addr : InetAddress.getAllByName(url.host())) {
+        if (!addr.isLoopbackAddress()) {
+          throw new IllegalArgumentException(url.host() + " is not loopback only host");
+        }
+      }
+    } catch (UnknownHostException e) {
+      throw new IllegalStateException("Host in " + url + " is not loopback address");
+    }
+  }
+
+  private Credentials fetchCredentials(HttpUrl url) {
+    try (Response response =
+        httpClient.newCall(new Request.Builder().url(url).method("GET", null).build()).execute()) {
+      if (!response.isSuccessful()) {
+        throw new IllegalStateException(url + " failed with HTTP status code " + response.code());
+      }
+
+      EcsCredentials creds = mapper.readValue(response.body().charStream(), EcsCredentials.class);
+      if (!"Success".equals(creds.code())) {
+        throw new IllegalStateException(url + " failed with message " + creds.message());
+      }
+      return creds.toCredentials();
+    } catch (IOException e) {
+      throw new IllegalStateException("Unable to parse response", e);
+    }
+  }
+
+  private Jwt getToken() {
+    String tokenFile = getProperty("AWS_WEB_IDENTITY_TOKEN_FILE");
+    try {
+      byte[] data = Files.readAllBytes(Paths.get(tokenFile));
+      return new Jwt(new String(data, StandardCharsets.UTF_8), 0);
+    } catch (IOException e) {
+      throw new IllegalStateException("Error in reading file " + tokenFile, e);
+    }
+  }
+
+  @Override
+  public synchronized Credentials fetch() {
+    if (credentials != null && !credentials.isExpired()) {
+      return credentials;
+    }
+
+    HttpUrl url = this.customEndpoint;
+    if (getProperty("AWS_WEB_IDENTITY_TOKEN_FILE") != null) {
```

Comment:
To making this call twice with same argument, we can save the output of this`getProperty` in a variable `tokenFile` and pass it as argument to `getToken`